### PR TITLE
Vector initialization improvements for vector forward mode AD

### PR DESCRIPTION
**Pull request by vaithak**
Improved initialization of vectors while differentiating variable declarations and added test cases to demonstrate vector mode support for programs with loops.

New ways of initializing `clad::array` is added by passing `(size, expression...)`. 
This is done to allow initializing the array by all constant values or expressions which evaluate to a single value; 
- for ex: `clad::array<double> arr(5UL, 1)` can be used to initialize a vector of size 5 with all entries as 1.

### CODE REVIEW

**Review by vgvassilev — COMMENTED** [review]
Can you move the first commit in a separate PR
?

**Comment by vaithak** [comment]
rebased with changes on master

File: docs/userDocs/source/user/CoreConcepts.rst [hunk 1]
```diff
@@ -4,12 +4,128 @@ Core Concepts
 How Clad Works
 =================
 
-.. todo:: todo
+Clad is an open source clang plugin that enables automatic differentiation of
```
Comment (parth-07) [inline docs/userDocs/source/user/CoreConcepts.rst:7]: This change should not be part of this pull request. Can you please check if you rebased the branch correctly?

**Review by parth-07 — APPROVED** [review]
Looks good.

File: docs/userDocs/source/user/CoreConcepts.rst (hunk 1 above)
Comment (vaithak) [reply docs/userDocs/source/user/CoreConcepts.rst]: There was a force push on the master which brought this somehow, rebased it properly 👍🏼

Comment (vgvassilev) [reply docs/userDocs/source/user/CoreConcepts.rst]: Yes, indeed and oversight in a PR merge on my end.

**Comment by vgvassilev** [comment]
Did you run clang-format?

**Comment by vaithak** [comment]
> Did you run clang-format?

yup 👍🏼 .

**Review by vgvassilev — APPROVED** [review]
